pkg/queue/metrics.go

```diff
@@ -1,93 +1,143 @@
 )
 
-// largest bucket is 5 seconds
-var durationMsBuckets = []float64{10, 50, 100, 200, 300, 500, 1000, 2000, 3000, 5000}
-
-var (
-	queueMetricLabels        = []string{"queue"}
-	taskMetricLabels         = []string{"queue", "type"}
-	oneTaskQueueMetricsSetup sync.Once
-	oneSchedulerMetricsSetup sync.Once
-)
-
 // TaskQueueMetricsType provides access to the prometheus metric objects for the task queue
 type TaskQueueMetricsType struct {
 	Labels          []string
 	TaskCounter     *prometheus.CounterVec
 	EnqueueDuration *prometheus.HistogramVec
 }
 
-// TaskQueueMetrics is the global metrics instance for the task queue of this instance
-var TaskQueueMetrics TaskQueueMetricsType
-
-// SetupTaskQueueMetrics must be called before any other call to the metric subsystem happens
-func SetupTaskQueueMetrics(namespace string) {
-	oneTaskQueueMetricsSetup.Do(func() {
-		TaskQueueMetrics = TaskQueueMetricsType{
-			Labels: queueMetricLabels,
-			TaskCounter: promauto.NewCounterVec(
-				prometheus.CounterOpts{
-					Namespace: namespace,
-					Subsystem: "queue",
-					Name:      "task",
-					Help:      "count of tasks that have been enqueued",
-				},
-				taskMetricLabels,
-			),
-			EnqueueDuration: promauto.NewHistogramVec(
-				prometheus.HistogramOpts{
-					Namespace: namespace,
-					Subsystem: "queue",
-					Name:      "enqueue_duration_ms",
-					Help:      "duration the enqueue action in ms",
-					Buckets:   durationMsBuckets,
-				},
-				queueMetricLabels,
-			),
-		}
-	})
-}
-
 // TaskQueueMetricsType provides access to the prometheus metric objects for the scheduler
 type SchedulerMetricsType struct {
 	Labels          []string
 	ScheduleCounter *prometheus.CounterVec
 	ErrorCounter    *prometheus.CounterVec
 }
 
-// SchedulerMetrics is the global metrics instance for the scheduler of this instance
-var SchedulerMetrics SchedulerMetricsType
-
-// SetupSchedulerMetrics must be called before any other call to the metric subsystem happens
-func SetupSchedulerMetrics(namespace string) {
-	oneSchedulerMetricsSetup.Do(func() {
-		SchedulerMetrics = SchedulerMetricsType{
-			Labels: queueMetricLabels,
-			ScheduleCounter: promauto.NewCounterVec(
-				prometheus.CounterOpts{
-					Namespace: namespace,
-					Subsystem: "scheduler",
-					Name:      "task",
-					Help:      "count of tasks that have been scheduled",
-				},
-				taskMetricLabels,
-			),
-			ErrorCounter: promauto.NewCounterVec(
-				prometheus.CounterOpts{
-					Namespace: namespace,
-					Subsystem: "scheduler",
-					Name:      "error",
-					Help:      "count of errors while scheduling",
-				},
-				taskMetricLabels,
-			),
+var (
+	// largest bucket is 5 seconds
+	durationMsBuckets  = []float64{10, 50, 100, 200, 300, 500, 1000, 2000, 3000, 5000}
+	queueMetricLabels  = []string{"queue"}
+	taskMetricLabels   = []string{"queue", "type"}
+	processName        = escapeNamespace(filepath.Base(os.Args[0]))
+	defTaskCounterOpts = prometheus.CounterOpts{
+		Namespace: processName,
```

Comment:
It is bettter to add the process name as a label and use the same metrics namespace in all implementations of the queue, that means when prometheus scrapes the metrics, we always see the same metrics and can easily subselect a specific component. Using a different namespace makes it harder to reuse the same dashboard for different systems.

Reply:
But I thought the whole point of the namespace was to separate by services, the rest of the metrics are split that way, aren't they?

Reply:
Done.